### PR TITLE
Retarget to .NET 5.0

### PR DESCRIPTION
**Pull request by RussKie**
## Proposed changes

1. Update dependencies with darc tool:     `darc update-dependencies --channel ".NET Core 5 Dev"`

2. Manually update references to `netcoreapp5`

3.  Delete all the hard-coded assembly versions and just rely on the product version since the version is higher than the NET Fx

4. Update serialization tests due to assembly version change.
Test serialisation only one way .NET Fx -> .NET, as well as .NET -> .NET. 
Serialisation .NET -> .NET Fx will need to be added later, if it is deemed necessary.

## Test methodology 

- manually build
  image

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/1684)

### CODE REVIEW

**Comment by RussKie** [comment]
I hit one snag though - I can only build from a VS or a VS developer prompt, otherwise the build is unable to locate ildasm.exe
image

Ideally I'd like the build to work from any cli, not just a VS developer prompt.
Any ideas on how to approach this correctly is appreciated.

**Comment by RussKie** [comment]
It is also failing on the ADO:
image

**Comment by AdamYoblick** [comment]
Looks good to me, but Eric and Will would know much better. I've never moved us up to a new major version sdk before :)

**Comment by ericstj** [comment]
Here's the relevant code: https://github.com/dotnet/coreclr/blob/4e960d72ad19086a9615229b24b50dbd9ccb143c/src/.nuget/Microsoft.NET.Sdk.IL/targets/Microsoft.NET.Sdk.IL.targets

For some reason you're no longer restoring ILDasm correctly.

**Comment by ericstj** [comment]
Could be related to https://github.com/dotnet/coreclr/commit/4e960d72ad19086a9615229b24b50dbd9ccb143c  /cc @jkoritzinsky

**Comment by jkoritzinsky** [comment]
There's a typo in that PR that I missed since it's only on the ildasm side. 😢 

I'll put out a PR to coreclr to fix it and ping back here once the fix is in.

**Comment by RussKie** [comment]
@wtgodbe - thank you; updated, please have a look

File: pkg/Microsoft.Dotnet.WinForms.ProjectTemplates/content/WinFormsLibrary-CSharp/Company.ClassLibrary1.csproj [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">
 
   <PropertyGroup>
-    <TargetFramework Condition="'$(TargetFrameworkOverride)' == ''">netcoreapp3.0</TargetFramework>
+    <TargetFramework Condition="'$(TargetFrameworkOverride)' == ''">netcoreapp5.0</TargetFramework>
```
[the remark this answers is not in the document]
Comment (RussKie) [reply pkg/Microsoft.Dotnet.WinForms.ProjectTemplates/content/WinFormsLibrary-CSharp/Company.ClassLibrary1.csproj]: These are templates and don't get build.
This is what a user gets running `dotnet new winforms`

**Comment by ericstj** [comment]
Do you want to delete all the hardcoded assembly versions now that you can just rely on the product version and be higher than desktop?
https://github.com/dotnet/winforms/search?q=AssemblyVersion&unscoped_q=AssemblyVersion
(similarly in 3.1 you might want to manually increment all these).

**Comment by RussKie** [comment]
> Do you want to delete all the hardcoded assembly versions now that you can just rely on the product version and be higher than desktop?

Thank you

> similarly in 3.1 you might want to manually increment all these

You've lost me here, increment to what?

File: src/System.Windows.Forms/Directory.Build.props [hunk 2]
```diff
@@ -2,8 +2,6 @@
 <Project>
   <Import Project="..\..\Directory.Build.props" />
   <PropertyGroup>
-    <!-- Identity needs to match what was in desktop so we remain compatible with desktop assemblies built against winforms -->
-    <AssemblyVersion>4.0.0.0</AssemblyVersion>
```
Comment (RussKie) [inline src/System.Windows.Forms/Directory.Build.props]: @ericstj as soon as I removed this line a number of tests started failing, e.g. https://dev.azure.com/dnceng/public/_build/results?buildId=330259&view=ms.vss-test-web.build-test-results-tab
image

What are we missing?

Comment (RussKie) [reply src/System.Windows.Forms/Directory.Build.props]: I guess the failure is attributed to the fact the assembly version is changed:

| Version | |
|---|-----|
| 3.0 | image |
| 5.0 | image |

Comment (RussKie) [reply src/System.Windows.Forms/Directory.Build.props]: Where does `42.42.42.42` come from though? I couldn't find it in the codebase

**Comment by ericstj** [comment]
> > similarly in 3.1 you might want to manually increment all these

> You've lost me here, increment to what?

In 3.1 you probably want a different assembly version for your assemblies.  In CoreFx we followed the pattern of incrementing 3rd portion by default on a new release (reserving 4th part for servicing), we'd increment the second/minor version if API changed.  Essentially doing our best to follow SemVer by communicating bugfixes in 3rd portion API/features in 2nd portion.  This is really up to you to decide.  After 3.x it's much easier since you can just rely on product versioning.

You may encounter some issues with the version bump since it will be the first time the .NETCore assembly versions differ from desktop for WinForms, but I think this is something you need to do to see what breaks.  Best to do it before all the design-time work is "done".

File: src/System.Windows.Forms/Directory.Build.props (hunk 2 above)
Comment (RussKie) [reply src/System.Windows.Forms/Directory.Build.props]: Thank you Jeremy.

**Comment by RussKie** [comment]
> In 3.1 you probably want a different assembly version for your assemblies. In CoreFx we followed the pattern of incrementing 3rd portion by default on a new release (reserving 4th part for servicing), we'd increment the second/minor version if API changed. Essentially doing our best to follow SemVer by communicating bugfixes in 3rd portion API/features in 2nd portion. This is really up to you to decide. After 3.x it's much easier since you can just rely on product versioning.

So something like `<AssemblyVersion>4.0.1.0</AssemblyVersion>`?

**Comment by zsd4yr** [comment]
> So something like `<AssemblyVersion>4.0.1.0</AssemblyVersion>`?

For 5.0, I think we want to start on 5.0.0.0 ? 

I think @ericstj 	is suggesting 4.0.1.0 for our 3.1 branch

**Comment by ericstj** [comment]
> So something like `<AssemblyVersion>4.0.1.0</AssemblyVersion>`

Yeah.  Whatever you choose as the policy it'd be good to document what you decide and when to change it.

**Comment by RussKie** [comment]
> > So something like `<AssemblyVersion>4.0.1.0</AssemblyVersion>`?
> 
> For 5.0, I think we want to start on 5.0.0.0 ?
> 
> I think @ericstj Eric St. John FTE is suggesting 4.0.1.0 for our 3.1 branch

Yes, that discussion was about 3.1 branch :)

File: src/System.Windows.Forms/Directory.Build.props (hunk 2 above)
Comment (RussKie) [reply src/System.Windows.Forms/Directory.Build.props]: I had an offline chat with Eric and I'm going to fix our serialisation tests to flow NET Fx --> NET5 for now.
@merriemcgaw @Tanya-Solyanik we need to decide how important NET5 --> NET Fx serialisation compat

File: Directory.Build.props [hunk 3]
```diff
@@ -5,6 +5,7 @@
   <Import Project="eng\FacadeAssemblies.props" />
 
   <PropertyGroup>
+    <TargetFramework>netcoreapp5.0</TargetFramework>
```
[the remark this answers is not in the document]
Comment (RussKie) [reply Directory.Build.props]: done

**Pull request merged**